cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -4504,39 +4500,203 @@ mod tests {
+            (
+                r#"permit(principal, action == ?resource, resource);"#,
+                ExpectedErrorMessage::error("expected single entity uid or set of entity uids, got: template slot"),
+            ),
+            (
+                r#"permit(principal, action in ?resource, resource);"#,
+                ExpectedErrorMessage::error("expected single entity uid or set of entity uids, got: template slot"),
+            ),
+            (
+                r#"permit(principal, action in [?bar], resource);"#,
+                ExpectedErrorMessage::error("expected single entity uid, got: template slot"),
             ),
         ];
 
         for (p_src, expected) in invalid_policies {
             assert_matches!(parse_policy_template(None, p_src), Err(e) => {
                 expect_err(p_src, &e, &expected);
             });
+            let forbid_src = format!("forbid{}", &p_src[6..]);
+            assert_matches!(parse_policy_template(None, &forbid_src), Err(e) => {
+                expect_err(forbid_src.as_str(), &e, &expected);
+            });
         }
     }
+
+    #[test]
+    fn missing_scope_constraint() {
+        let p_src = "permit();";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error("this policy is missing the `principal` variable in the scope"));
+        });
+        let p_src = "permit(principal);";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error("this policy is missing the `action` variable in the scope"));
+        });
+        let p_src = "permit(principal, action);";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error("this policy is missing the `resource` variable in the scope"));
+        });
+    }
+
+    #[test]
+    fn invalid_scope_constraint() {
+        let p_src = "permit(foo, action, resource);";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error_and_help(
+                "expected a variable that is valid in the policy scope; found: `foo`",
+                "must be one of `principal`, `action`, or `resource`"
+            ));
+        });
+        let p_src = "permit(resource, action, resource);";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error("the variable `resource` is invalid in this policy scope clause, the variable `principal` is expected"));
+        });
+
+        let p_src = "permit(principal, principal, resource);";
+        assert_matches!(parse_policy_template(None, p_src), Err(e) => {
+            expect_err(p_src, &e, &ExpectedErrorMessage::error("the variable `principal` is invalid in this policy scope clause, the variable `action` is expected"));
```

Comment:
this error message isn't the best -- `principal` isn't invalid in a policy scope clause.  I'll make an issue

Reply:
Technically correct that it's invalid in _this_ clause. It makes more sense when the source range is rendered. It can absolutely be improved though.

Reply:
#562